File: OptiPlanBackend/OptiPlanBackend/Models/WorkItem.cs

using OptiPlanBackend.Enums;
using System.ComponentModel.DataAnnotations;

namespace OptiPlanBackend.Models
{
    public class WorkItem
    {
        public Guid Id { get; set; }
        public string Title { get; set; } = string.Empty;
        public string? Description { get; set; }

        public WorkItemType Type { get; set; }  // Story, Bug, Task, Epic, Subtask
        public WorkItemStatus Status { get; set; }
        public WorkItemPriority Priority { get; set; }

        public Guid ProjectId { get; set; }
        public Project Project { get; set; }

        public Guid? SprintId { get; set; }
        public Sprint? Sprint { get; set; }

        public Guid? ParentId { get; set; }  // for nesting (Story → Subtask)
        public WorkItem? Parent { get; set; }

        public ICollection<WorkItem> SubItems { get; set; } = new List<WorkItem>(); // e.g. subtasks

        public Guid? AssignedUserId { get; set; }
        public User? AssignedUser { get; set; }

        public Guid? ReporterId { get; set; }
        public User? Reporter { get; set; }

        public DateTime? StartDate { get; set; }
        public int? EstimatedHours { get; set; }
        public string? Labels { get; set; } 



        public ICollection<Comment> Comments { get; set; } = new List<Comment>();
        public ICollection<Attachment> Attachments { get; set; } = new List<Attachment>();

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? DueDate { get; set; }
        public DateTime? CompletedAt { get; set; }

        public int? StoryPoints { get; set; }  // Only for Story-type
        public double CompletionPercentage { get; set; } = 0;

        public bool IsBlocked { get; set; }
        public string? BlockReason { get; set; }

        public bool IsCompleted => Status == WorkItemStatus.Done;


        public ICollection<WorkItemHistory> History { get; set; } = new List<WorkItemHistory>();
    }

}

File: optiplan-frontend/src/app/models/work-item.ts
import { Attachment } from "./attachmen";
import { Comment } from "./comment";
import { WorkItemPriority } from "./enums/work-item-priority";
import { WorkItemStatus } from "./enums/work-item-status";
import { WorkItemType } from "./enums/work-item-type";
import { User } from "./user";
import { WorkItemHistory } from "./work-item-history";


export interface WorkItem {
  updatedAt: string | number | Date;
  id: string;
  title: string;
  description?: string;

  type: WorkItemType;
  status: WorkItemStatus;
  priority: WorkItemPriority;

  projectId: string;
  sprintId?: string;

  parentId?: string;
  parent?: WorkItem;

  subItems?: WorkItem[];

  assignedUserId?: string;
  assignedUser?: User;

  reporterId?: string;
  reporter?: User;

  startDate?: Date;
  estimatedHours?: number;
  labels?: string;

  comments?: Comment  [];
  attachments?: Attachment[];

  createdAt: Date;
  dueDate?: Date;
  completedAt?: Date;

  storyPoints?: number;
  completionPercentage: number;

  isBlocked: boolean;
  blockReason?: string;

  // Calculated in backend
  isCompleted: boolean;

  history?: WorkItemHistory[];  // Define this interface if not already done
}
